esx_service/utils/vm_listener.py

```diff
@@ -0,0 +1,154 @@
+        logging.error(err_msg)
+        return err_msg
+
+
+def listen_vm_propertychange(pc):
+    """
+    Waits for updates on powerstate of VMs. If powerstate is poweroff,
+    detach the dvs managed volumes attached to VM
+    """
+    logging.info("VMChangeListener thread started")
+    version = ''
+    while True:
+        result = pc.WaitForUpdates(version)
+        try:
+            # process the updates result
+            for filterSet in result.filterSet:
+                for objectSet in filterSet.objectSet:
+                    if objectSet.kind != 'modify':
+                        continue
+                    for change in objectSet.changeSet:
+                        # if the event was powerOff for a VM, set the status of all
+                        # docker volumes attached to the VM to be detached
+                        if change.name != VM_POWERSTATE or change.val != POWERSTATE_POWEROFF:
+                            continue
+
+                        moref = getattr(objectSet, 'obj', None)
+                        # Do we need to alert the admin? how?
+                        if not moref:
+                            logging.error("Could not retrieve the managed object")
+                            continue
+
+                        logging.info("VM poweroff change found for %s", moref.config.name)
+
+                        set_device_detached(moref)
+        except Exception as e:
+            # Do we need to alert the admin? how?
+            logging.error("VMChangeListener: error %s", str(e))
+
+        version = result.version
+
+    logging.info("VMChangeListener thread exiting")
+
+
+def vm_folder_traversal():
+    """
+    Build the traversal spec for the property collector to traverse vmFolder
+    """
+
+    TraversalSpec = vmodl.query.PropertyCollector.TraversalSpec
+    SelectionSpec = vmodl.query.PropertyCollector.SelectionSpec
+
+    # Traversal through vmFolder branch
+    dcToVmf = TraversalSpec(name='dcToVmf', type=vim.Datacenter, path='vmFolder', skip=False)
+    dcToVmf.selectSet.append(SelectionSpec(name='visitFolders'))
+
+    # Recurse through the folders
+    visitFolders = TraversalSpec(name='visitFolders', type=vim.Folder, path='childEntity', skip=False)
+    visitFolders.selectSet.extend((SelectionSpec(name='visitFolders'), SelectionSpec(name='dcToVmf'),))
+
+    return SelectionSpec.Array((visitFolders, dcToVmf,))
+
+
+def set_device_detached(vm_moref):
+    """
+    For all devices in device_list, if it is a DVS volume, set its status to detached in KV
+    """
+
+    for dev in vm_moref.config.hardware.device:
+        # if it is a dvs managed volume, set its status as detached
+        vmdk_path = vmdk_utils.find_dvs_volume(dev)
+        if vmdk_path:
+            logging.info("Setting detach status for %s", vmdk_path)
+            # disk detach and update the status in KV
+            err_msg = vmdk_ops.disk_detach_int(vmdk_path, vm_moref, dev)
+            if err_msg:
```

Comment:
Pls. also review the use of handle_stale_attach in the vmdk_ops.py and check if what its doing is relevant anymore.